pkg/trace/stats/quantile/ddsketch_converter.go

```diff
@@ -0,0 +1,180 @@
+
+func getIndexes(s1 ddSketch, s2 ddSketch) []int {
+	// todo: No need to re-allocate that array at each conversion.
+	// but this function needs to be thread safe in the agent.
+	indexes := make([]int, 0, s1.maxSize() + s2.maxSize())
+	for i := range s1.contiguousBins {
+		indexes = append(indexes, i + s1.offset)
+	}
+	for i := range s2.contiguousBins {
+		ind := i + s2.offset
+		if ind >= s1.offset && ind < s1.offset + len(s1.contiguousBins) {
+			continue
+		}
+		indexes = append(indexes, ind)
+	}
+	for i := range s1.bins {
+		ind := int(i)
+		if ind >= s1.offset && ind < s1.offset + len(s1.contiguousBins) {
+			continue
+		}
+		if ind >= s2.offset && ind < s2.offset + len(s2.contiguousBins) {
+			continue
+		}
+		indexes = append(indexes, ind)
+	}
+	for i := range s2.bins {
+		ind := int(i)
+		if ind >= s1.offset && ind < s1.offset + len(s1.contiguousBins) {
+			continue
+		}
+		if ind >= s2.offset && ind < s2.offset + len(s2.contiguousBins) {
+			continue
+		}
+		if _, ok := s1.bins[i]; ok {
+			continue
+		}
+		indexes = append(indexes, ind)
+	}
+	sort.Ints(indexes)
+	return indexes
+}
+
+// decodeDDSketch decodes a ddSketch from a protobuf encoded ddSketch
+// it only supports positive contiguous bins
+func decodeDDSketch(data []byte) (ddSketch, error) {
+	var sketchPb pb.DDSketch
+	if err := proto.Unmarshal(data, &sketchPb); err != nil {
+		return ddSketch{}, err
+	}
+	mapping, err := ddSketchMappingFromProto(sketchPb.Mapping)
+	if err != nil {
+		return ddSketch{}, err
+	}
+	if sketchPb.Mapping.IndexOffset > 0 { err = errors.New("index offset non 0") }
+	if  len(sketchPb.NegativeValues.BinCounts)> 0 { err = errors.New("contains negative values") }
+	if  len(sketchPb.NegativeValues.ContiguousBinCounts)> 0 { err = errors.New("contains negative values") }
+	if err != nil {
+		return ddSketch{}, errors.New("ddSketch format not supported: " + err.Error())
+	}
+	return ddSketch{
+		mapping: mapping,
+		bins: sketchPb.PositiveValues.BinCounts,
+		contiguousBins: sketchPb.PositiveValues.ContiguousBinCounts,
+		offset: int(sketchPb.PositiveValues.ContiguousBinIndexOffset),
+		zeros: int(sketchPb.ZeroCount),
+	}, nil
+}
+
+func ddSketchMappingFromProto(mappingPb *pb.IndexMapping) (m mapping.IndexMapping, err error) {
+	switch mappingPb.Interpolation {
+	case pb.IndexMapping_NONE:
+		return mapping.NewLogarithmicMappingWithGamma(mappingPb.Gamma, mappingPb.IndexOffset)
+	case pb.IndexMapping_LINEAR:
+		return mapping.NewLinearlyInterpolatedMappingWithGamma(mappingPb.Gamma, mappingPb.IndexOffset)
+	case pb.IndexMapping_CUBIC:
+		return mapping.NewCubicallyInterpolatedMappingWithGamma(mappingPb.Gamma, mappingPb.IndexOffset)
+	default:
+		return nil, fmt.Errorf("interpolation not supported: %d", mappingPb.Interpolation)
+	}
+}
+
+// DDToGKSketches converts two dd sketches: ok and errors to 2 gk sketches: hits and errors
+// with hits = ok + errors
+func DDToGKSketches(okSketchData []byte, errSketchData []byte) (hits, errors *SliceSummary, err error) {
```

Comment:
Can we have a single func called `DDSketchToGK` which takes only one argument `[]byte` and converts it? It makes more sense imho from an API perspective.

I'm not very familiar with the code, so if they are needed together, ignore me :)

Reply:
That would be ideal yes. But hits = ok + errors, so we need to sum them at the same time as converting them, hence the function that take the two.
An alternative would be to:
1. Decode to ddsketch
2. Sum ok and errors
3. Convert hits and errors to gksketch

But that leads to more allocations (to sum ok and errors)

Reply:
All good then.